File: Locations/Location.java
package Locations;

import People.Person;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Mark Chimes on 2017/11/05.
 */
public abstract class Location {
    protected List<Person> peopleHere = new ArrayList<>();

    public abstract List<String> description();
    public abstract String name();
    public abstract List<String> nextMessage();

    public List<Location> getSublocations() {
        return new ArrayList<>();
    }

    public List<Person> getPeopleRecursive() {
        if (getSublocations().isEmpty()) {
            return peopleHere;
        }

        List<Person> people = new ArrayList<>();

        for (Location subLocation : getSublocations()) {
            people.addAll(subLocation.getPeopleRecursive());
        }

        return people;
    }
}

File: Locations/Location.kt
package Locations

import People.Person

import java.util.ArrayList

/**
 * Created by Mark Chimes on 2017/11/05.
 */
abstract class Location(protected var allPeople: List<Person>, val sublocations: List<Location>) {

    abstract fun description(): List<String>
    abstract fun name(): String
    abstract fun nextMessage(): List<String>

    val peopleRecursive: List<Person>
        get() {
            val peopleHere = ArrayList<Person>()

            if (sublocations.isEmpty()) {
                for (person in allPeople) {
                    if (person.currentLocation() == this) {
                        peopleHere.add(person)
                    }
                }

                return peopleHere
            }

            for (subLocation in sublocations) {
                peopleHere.addAll(subLocation.peopleRecursive)
            }

            return peopleHere
        }
}
